server/pulp/server/db/model/__init__.py

```diff
@@ -307,3 +312,132 @@ def post_save(cls, sender, document, **kwargs):
+    :type storage_path: mongoengine.StringField
+
+    :ivar _ns: The namespace field (Deprecated), reading
+    :type _ns: mongoengine.StringField
+    :ivar unit_type_id: string representation of content unit type
+    :type unit_type_id: mongoengine.StringField
+    """
+
+    id = StringField(primary_key=True)
+    last_updated = IntField(db_field='_last_updated', required=True)
+    user_metadata = DictField(db_field='pulp_user_metadata')
+    storage_path = StringField(db_field='_storage_path')
+
+    # For backward compatibility
+    _ns = StringField(required=True)
+    unit_type_id = StringField(db_field='_content_type_id', required=True)
+    _source_location = None
+    _relative_path = None
+
+    meta = {
+        'abstract': True,
+    }
+
+    @classmethod
+    def post_init_signal(cls, sender, document):
+        """
+        The signal that is triggered before a unit is initialized
+
+        This is used to validate that the unit_key_fields attribute is set properly
+
+        :param sender: sender class
+        :type sender: object
+        :param document: Document that sent the signal
+        :type document: ContentUnit
+        """
+
+        if not hasattr(document, 'unit_key_fields'):
+            class_name = type(document).__name__
+            raise exceptions.PulpCodedException(error_codes.PLP0035, class_name=class_name)
+
+    @classmethod
+    def pre_save_signal(cls, sender, document, **kwargs):
+        """
+        The signal that is triggered before a unit is saved
+
+        :param sender: sender class
+        :type sender: object
+        :param document: Document that sent the signal
+        :type document: ContentUnit
+        """
+        if not document.id:
+            document.id = str(uuid.uuid4())
+        document.last_updated = dateutils.now_utc_timestamp()
+
+    @classmethod
+    def post_save_signal(cls, sender, document, **kwargs):
+        """
+        The signal that is triggered after a unit is saved
+
+        :param sender: sender class
+        :type sender: object
+        :param document: Document that sent the signal
+        :type document: ContentUnit
+        """
+        # If content was set on this unit, move the content into place
+        if document._source_location and document._relative_path:
+            server_storage_dir = config.config.get('server', 'storage_dir')
+            platform_storage_dir = os.path.join(server_storage_dir, 'content',
+                                                document.unit_type_id)
+            target_location = os.path.join(platform_storage_dir, document._relative_path)
+            # Make sure the directory for the target exists
+            if os.path.isdir(document._source_location):
+                shutil.copytree(document._source_location, target_location)
+            else:
+                check_dir = os.path.dirname(target_location)
+                misc.mkdir(check_dir)
```

Comment:
I think having a try/except here would be better than using the `misc.mkdir()`. Something [just like this](https://github.com/pulp/pulp/pull/1809/files#diff-83c54e49104a094800b69d01f7385232R292) would make the behavior of this code clearer and more intentional.

The reason I don't want us to use `misc.mkdir` is because it is very similarly named to the standard library function, but with a completely different behavior. This causes confusion and the line savings aren't worth it.